Add progressiveGAN for 3D brain MR images

wazeerzulfikar opened the pull request:
## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Summary
<!--- What does your code do? -->
Adds a generative adversarial network which uses a progressively growing generator and discriminator to generate high resolution brain MR images. Trained using a wasserstein loss with gradient penalty.

## Checklist
<!--- Please, let us know if you need help-->
- [ ] All tests passing
- [x] I have added tests to cover my changes
- [x] I have updated documentation (if necessary)
- [x] My code follows the code style of this project
(we are using `black`: you can `pip install pre-commit`,
run `pre-commit install` in the `pydra` directory
and `black` will be run automatically with each commit)

## TODO
- [x] Add networks
- [x] Add losses for training
- [x] Add multi-resolution tfrecord writing
- [x] Add trainer component
- [x] Add example guide to run training
- [x] Add inference using saved model
- [x] Add cli for generation
- [x] Add tests for the networks
- [ ] Add visual evaluation tools (interpolation, nearest neighbor etc)

## Acknowledgment
- [x] I acknowledge that this contribution will be available under the Apache 2 license.

Aakanksha-Rana commented:
Reviewing!

wazeerzulfikar commented:
@satra what sections would we want to update the README with?
I was thinking of the following :
- A nobrainer CLI example to use a pre-trained network to generate images
- A visual image generated in progressive resolutions
- Maybe a few more visualizations like the interpolation 

What do you suggest?

satra commented:
for the pre-trained network we should add it to trained-models repo (@Hoda1394 can help). and then let's generate a google collab notebook to demonstrate. 

in the readme, i would point them to the collab to try things out.

wazeerzulfikar commented:
Thanks, what about a visual example? and the CLI?

for instance, 
```
docker run --rm -v $PWD:/data neuronets/nobrainer \
  generate \
    --model=/models/saved_models \
    --target_shape 128 128 128 \
    /data/generated.nii.gz
```

satra commented:
a CLI option would be nice - but i would want the model to be pulled from trained-models. so we may need to call it used the path in the trained-models repo.

regarding an image, we can add to the README, but i was hoping to reorganize the readme and put details for each model in a separate location. we should probably start generating docs and adding such things to the docs.

Hoda1394 commented:
I will also review the code, so the PR can be merged after two reviews.

wazeerzulfikar commented:
Thank you @Aakanksha-Rana and @Hoda1394 for reviewing it!

@Hoda1394 , there are a set of saved models, as it is multiple networks at different resolutions. I have a new set training now (with the updated savedmodel format), but for now, can we use the ones in `/scratch/user/wazeer/models/progressivegan` on openmind?

Is it easy to update them once it's uploaded for the first time?

Hoda1394 commented:
@wazeerzulfikar- actually, it is not that hard to update it. I think it is better to update the model when you think a version is ready to be released and maybe it is significantly different with the previous version.
@satra -what do you think?

satra commented:
it's ok to put something an update it. the main thing is to indicate how to use the model.

wazeerzulfikar commented:
I added images of the latest generated brain scans to the `neuronets/trained-models/images` through this PR https://github.com/neuronets/trained-models/pull/9.
These images are used in the updated README of `nobrainer`.

@Hoda1394 I have also updated the models at `/scratch/user/wazeer/models/progressivegan`, you can add them to trained-models repo

satra commented on README.md:
````diff
@@ -126,6 +129,33 @@ docker run --rm -v $PWD:/data neuronets/nobrainer \
+```bash
+docker run --rm -v $PWD:/data neuronets/nobrainer \
+  generate \
+    --model=/models/progressivegan_models \
````
i think this readme needs to be updated (in general) for all the examples to work. i'm not sure any examples work right now. @Hoda1394  any thoughts? this would also mean addition of the models to trained models, updating the docker generator to pull in the trained-models and then updating this readme.

Hoda1394 replied:
I will check the examples to make sure that they are working as they are. but we need to change the readme and add the instructions for pulling the model and running the examples.

satra commented:
@wazeerzulfikar  - the main thing this PR needs are tests for all the functions that were added. i couldn't see a single test. and also a basic test for the network itself. it doesn't have to produce the write thing, but it should run and save models.

wazeerzulfikar commented:
@satra you are right, I will add the tests and update the PR. Thanks!

wazeerzulfikar commented:
Added the tests

wazeerzulfikar commented:
Is there way to trigger the Travis CI for the running the tests? It is not seeming to run automatically for this PR
